### ISSUE TITLE
Question Regarding O2M Implementation in RT-DETRv3

### ISSUE THREAD

**Issue by Anchor1566** [clxia12/RT-DETRv3#3]
First of all, thank you so much for your amazing work on RT-DETR. I’ve been following your work and am particularly interested in RT-DETRv3 and its efficient framework design, which leverages Hungarian matching and the novel hierarchical dense positive supervision method. 
I am currently working on implementing Group DETR based on RT-DETR, and I noticed that in your code, specifically in `group_detr_transformer.py`, you have the implementation of the `GroupDINOTransformer(nn.Layer)` class. I’ve been experimenting with this idea over the past week or two, but unfortunately, my implementation is underperforming, and I’ve observed a significant drop in $AP_{50}$—around 4 AP.
Since there isn't an official implementation for Group DETR , I am unsure if I might have missed any critical details or steps. I would greatly appreciate it if you could provide any guidance on potential issues or mistakes that could cause such a large performance drop.

I keep the loss weight as the original RTDETR
Here is the key code:
## 1.decoder
```
class TransformerDecoderLayer(nn.Module):
def forward(self,
                target,
                reference_points,
                memory,
                memory_spatial_shapes,
                attn_mask=None,
                memory_mask=None,
                query_pos_embed=None, 
                num_groups=None):
        # self attention
        if self.training and num_groups is not None:
            bs, _, d = target.shape
            target = target.reshape(bs*num_groups, -1, d)
            query_pos_embed = query_pos_embed.reshape(bs*num_groups, -1, d)
            
        q = k = self.with_pos_embed(target, query_pos_embed)

        target2, _ = self.self_attn(q, k, value=target, attn_mask=attn_mask)
        target = target + self.dropout1(target2)
        target = self.norm1(target)

        if self.training and num_groups is not None:
            target = target.reshape(bs, -1, d)
            query_pos_embed = query_pos_embed.reshape(bs, -1, d)
        # cross attention
        target2 = self.cross_attn(\
            self.with_pos_embed(target, query_pos_embed), 
            reference_points, 
            memory, 
            memory_spatial_shapes, 
            memory_mask)
        target = target + self.dropout2(target2)
        target = self.norm2(target)

        # ffn
        target2 = self.forward_ffn(target)
        target = target + self.dropout4(target2)
        target = self.norm3(target)

        return target
        
def _get_decoder_input(self,
                           memory: torch.Tensor,
                           spatial_shapes,
                           denoising_logits=None,
                           denoising_bbox_unact=None):

        if self.training or self.eval_spatial_size is None:
            anchors, valid_mask = self._generate_anchors(spatial_shapes, device=memory.device)
        else:
            anchors = self.anchors
            valid_mask = self.valid_mask

        memory = valid_mask.to(memory.dtype) * memory  
        contents, enc_topk_bbox_unacts, enc_topk_bboxes_list, enc_topk_logits_list = [], [], [], []
        for g_id in range(self.num_groups):
            output_memory :torch.Tensor = self.enc_output[g_id](memory)
            enc_outputs_logits :torch.Tensor = self.enc_score_head[g_id](output_memory)
            enc_outputs_coord_unact :torch.Tensor = self.enc_bbox_head[g_id](output_memory) + anchors

            enc_topk_memory, enc_topk_logits, enc_topk_bbox_unact = \
                self._select_topk(output_memory, enc_outputs_logits, enc_outputs_coord_unact, self.num_queries)

            if self.training:
                enc_topk_bboxes = F.sigmoid(enc_topk_bbox_unact)
                enc_topk_bboxes_list.append(enc_topk_bboxes)
                enc_topk_logits_list.append(enc_topk_logits)

            if self.learn_query_content:
                content = self.tgt_embed.weight.unsqueeze(0).tile([memory.shape[0], 1, 1])
            else:
                content = enc_topk_memory.detach()

            enc_topk_bbox_unact = enc_topk_bbox_unact.detach()

            if denoising_bbox_unact is not None:
                enc_topk_bbox_unact = torch.concat([denoising_bbox_unact[g_id], enc_topk_bbox_unact], dim=1)
                content = torch.concat([denoising_logits[g_id], enc_topk_memory], dim=1)

            if not self.training:
                return content, enc_topk_bbox_unact, enc_topk_bboxes_list, enc_topk_logits_list

            contents.append(content)
            enc_topk_bbox_unacts.append(enc_topk_bbox_unact)

        contents = torch.concat(contents, dim=1)
        enc_topk_bbox_unacts = torch.concat(enc_topk_bbox_unacts, dim=1)
        enc_topk_bboxes_list = [torch.concat(enc_topk_bboxes_list, dim=1)]
        enc_topk_logits_list = [torch.concat(enc_topk_logits_list, dim=1)]

        return contents, enc_topk_bbox_unacts, enc_topk_bboxes_list, enc_topk_logits_list

def forward(self, feats, targets=None):
        # input projection and embedding
        # memory: [bs, len, hidden_dim]
        memory, spatial_shapes = self._get_encoder_input(feats)
        bs, _, _ = memory.shape

        # prepare denoising training
        if self.training and self.num_denoising > 0:
            denoising_logits_list = []
            denoising_bbox_unact_list = []
            for g_id in range(self.num_groups):
                denoising_logits, denoising_bbox_unact, attn_mask, dn_meta = \
                get_contrastive_denoising_training_group(targets, \
                    self.num_classes, 
                    self.num_queries, 
                    self.denoising_class_embed[g_id], 
                    num_denoising=self.num_denoising, 
                    label_noise_ratio=self.label_noise_ratio, 
                    box_noise_scale=self.box_noise_scale, 
                    num_groups=self.num_groups)
                denoising_logits_list.append(denoising_logits)
                denoising_bbox_unact_list.append(denoising_bbox_unact)

            denoising_logits = torch.stack(denoising_logits_list)
            denoising_bbox_unact = torch.stack(denoising_bbox_unact_list)

        else:
            denoising_logits, denoising_bbox_unact, attn_mask, dn_meta = None, None, None, None

        init_ref_contents, init_ref_points_unact, enc_topk_bboxes_list, enc_topk_logits_list = \
            self._get_decoder_input(memory, spatial_shapes, denoising_logits, denoising_bbox_unact)

        if self.training and self.num_groups is not None:
            init_ref_contents = init_ref_contents.reshape(bs*self.num_groups, -1, self.hidden_dim)
            init_ref_points_unact = init_ref_points_unact.reshape(bs*self.num_groups, -1, 4)
            dn_content, init_ref_content = torch.split(init_ref_contents, dn_meta['dn_num_split'], dim=1)
            dn_ref_points_unact, init_ref_points_unact = torch.split(init_ref_points_unact, dn_meta['dn_num_split'], dim=1)
            init_ref_contents = torch.cat([dn_content.reshape(bs, -1, dn_content.shape[-1]), init_ref_content.reshape(bs, -1, init_ref_content.shape[-1])], dim=1) 
            init_ref_points_unact = torch.cat([dn_ref_points_unact.reshape(bs, -1, dn_ref_points_unact.shape[-1]), init_ref_points_unact.reshape(bs, -1, init_ref_points_unact.shape[-1])], dim=1)

        # decoder
        # [num_layers, bs, num_queries, num_classes]
        out_bboxes, out_logits = self.decoder(
            init_ref_contents,
            init_ref_points_unact,
            memory,
            spatial_shapes,
            self.dec_bbox_head,
            self.dec_score_head,
            self.query_pos_head,
            attn_mask=attn_mask, 
            num_groups=self.num_groups)

        if self.training and dn_meta is not None and self.num_groups is not None:
            dn_out_bboxes, out_bboxes = torch.split(out_bboxes, dn_meta['dn_num_split_groups'], dim=2)
            dn_out_logits, out_logits = torch.split(out_logits, dn_meta['dn_num_split_groups'], dim=2)

        out = {'pred_logits': out_logits[-1], 'pred_boxes': out_bboxes[-1]}

        if self.training and self.aux_loss:
            out['aux_outputs'] = self._set_aux_loss(out_logits[:-1], out_bboxes[:-1])
            out['enc_aux_outputs'] = self._set_aux_loss(enc_topk_logits_list, enc_topk_bboxes_list)
            out['enc_meta'] = {'class_agnostic': self.query_select_method == 'agnostic'}

            if dn_meta is not None:
                out['dn_aux_outputs'] = self._set_aux_loss(dn_out_logits, dn_out_bboxes)
                out['dn_meta'] = dn_meta

        return out
 ``` 
        
## 2.matcher

        ```
        def forward(self, outputs: Dict[str, torch.Tensor], targets, num_groups):
        bs, num_queries = outputs["pred_logits"].shape[:2]


        # We flatten to compute the cost matrices in a batch
        if self.use_focal_loss:
            out_prob = F.sigmoid(outputs["pred_logits"].flatten(0, 1))
        else:
            out_prob = outputs["pred_logits"].flatten(0, 1).softmax(-1)  # [batch_size * num_queries, num_classes]

        out_bbox = outputs["pred_boxes"].flatten(0, 1)  # [batch_size * num_queries, 4]

        # Also concat the target labels and boxes
        tgt_ids = torch.cat([v["labels"] for v in targets])
        tgt_bbox = torch.cat([v["boxes"] for v in targets])

        # Compute the classification cost. Contrary to the loss, we don't use the NLL,
        # but approximate it in 1 - proba[target class].
        # The 1 is a constant that doesn't change the matching, it can be ommitted.
        if self.use_focal_loss:
            out_prob = out_prob[:, tgt_ids]
            neg_cost_class = (1 - self.alpha) * (out_prob ** self.gamma) * (-(1 - out_prob + 1e-8).log())
            pos_cost_class = self.alpha * ((1 - out_prob) ** self.gamma) * (-(out_prob + 1e-8).log())
            cost_class = pos_cost_class - neg_cost_class        
        else:
            cost_class = -out_prob[:, tgt_ids]

        # Compute the L1 cost between boxes
        cost_bbox = torch.cdist(out_bbox, tgt_bbox, p=1)

        # Compute the giou cost betwen boxes
        cost_giou = -generalized_box_iou(box_cxcywh_to_xyxy(out_bbox), box_cxcywh_to_xyxy(tgt_bbox))
        
        # Final cost matrix
        C = self.cost_bbox * cost_bbox + self.cost_class * cost_class + self.cost_giou * cost_giou
        C = C.view(bs, num_queries, -1).cpu()

        # for group matching
        sizes = [len(v["boxes"]) for v in targets]
        g_num_queries = num_queries // num_groups
        C_list = C.split(g_num_queries, dim=1)
        for g_i in range(num_groups):
            C_g = C_list[g_i]
            indices_g = [linear_sum_assignment(c[i]) for i, c in enumerate(C_g.split(sizes, -1))]
            if g_i == 0:
                indices = indices_g
            else:
                indices = [
                    (
                        np.concatenate([indice1[0], indice2[0] + g_num_queries * g_i]),
                        np.concatenate([indice1[1], indice2[1]]),
                    )
                    for indice1, indice2 in zip(indices, indices_g)
                ]
        indices = [(torch.as_tensor(i, dtype=torch.int64), torch.as_tensor(j, dtype=torch.int64)) for i, j in indices]
        return {'indices': indices}
        ```
        
## 3.criterion

        ```
        def forward(self, outputs, targets, **kwargs):
        num_groups = self.num_groups if self.training else 1
        outputs_without_aux = {k: v for k, v in outputs.items() if 'aux' not in k}

        # Compute the average number of target boxes accross all nodes, for normalization purposes
        num_boxes = sum(len(t["labels"]) for t in targets) * num_groups
        num_boxes = torch.as_tensor([num_boxes], dtype=torch.float, device=next(iter(outputs.values())).device)
        if is_dist_available_and_initialized():
            torch.distributed.all_reduce(num_boxes)
        num_boxes = torch.clamp(num_boxes / get_world_size(), min=1).item()
        
        # Retrieve the matching between the outputs of the last layer and the targets
        matched = self.matcher(outputs_without_aux, targets, num_groups)
        indices = matched['indices']

        # Compute all the requested losses
        losses = {}
        for loss in self.losses:
            meta = self.get_loss_meta_info(loss, outputs, targets, indices)            
            l_dict = self.get_loss(loss, outputs, targets, indices, num_boxes, **meta)
            l_dict = {k: l_dict[k] * self.weight_dict[k] for k in l_dict if k in self.weight_dict}
            losses.update(l_dict)

        # In case of auxiliary losses, we repeat this process with the output of each intermediate layer.
        if 'aux_outputs' in outputs:
            for i, aux_outputs in enumerate(outputs['aux_outputs']):
                if not self.share_matched_indices:
                    matched = self.matcher(aux_outputs, targets, num_groups)
                    indices = matched['indices']
                for loss in self.losses:
                    meta = self.get_loss_meta_info(loss, aux_outputs, targets, indices)
                    l_dict = self.get_loss(loss, aux_outputs, targets, indices, num_boxes, **meta)
                    l_dict = {k: l_dict[k] * self.weight_dict[k] for k in l_dict if k in self.weight_dict}
                    l_dict = {k + f'_aux_{i}': v for k, v in l_dict.items()}
                    losses.update(l_dict)

        # In case of cdn auxiliary losses. For rtdetr
        if 'dn_aux_outputs' in outputs:
            assert 'dn_meta' in outputs, ''
            indices = self.get_cdn_matched_indices(outputs['dn_meta'], targets)
            dn_num_boxes = num_boxes * outputs['dn_meta']['dn_num_group']
            for i, aux_outputs in enumerate(outputs['dn_aux_outputs']):
                for loss in self.losses:
                    meta = self.get_loss_meta_info(loss, aux_outputs, targets, indices)
                    l_dict = self.get_loss(loss, aux_outputs, targets, indices, dn_num_boxes, **meta)
                    l_dict = {k: l_dict[k] * self.weight_dict[k] for k in l_dict if k in self.weight_dict}
                    l_dict = {k + f'_dn_{i}': v for k, v in l_dict.items()}
                    losses.update(l_dict)

        # In case of encoder auxiliary losses. For rtdetr v2
        if 'enc_aux_outputs' in outputs:
            assert 'enc_meta' in outputs, ''
            class_agnostic = outputs['enc_meta']['class_agnostic']
            if class_agnostic:
                orig_num_classes = self.num_classes
                self.num_classes = 1
                enc_targets = copy.deepcopy(targets)
                for t in enc_targets:
                    t['labels'] = torch.zeros_like(t["labels"])
            else:
                enc_targets = targets

            for i, aux_outputs in enumerate(outputs['enc_aux_outputs']):
                matched = self.matcher(aux_outputs, targets, num_groups)
                indices = matched['indices']
                for loss in self.losses:
                    meta = self.get_loss_meta_info(loss, aux_outputs, enc_targets, indices)
                    l_dict = self.get_loss(loss, aux_outputs, enc_targets, indices, num_boxes, **meta)
                    l_dict = {k: l_dict[k] * self.weight_dict[k] for k in l_dict if k in self.weight_dict}
                    l_dict = {k + f'_enc_{i}': v for k, v in l_dict.items()}
                    losses.update(l_dict)
            
            if class_agnostic:
                self.num_classes = orig_num_classes

        return losses
        ```

Thank you again for your invaluable work and contributions to the community. I look forward to your insights, and I wish you continued success both in your academic and personal endeavors!
